openid-4-verifiable-presentations-1_0.md

```diff
@@ -432,54 +427,43 @@ Note: "https://self-issued.me/v2" is a symbolic string and can be used as an `au
 
-The following is a non-normative example of a header and a body of a signed Request Object when Client Identifier scheme is a `did`:
+Existing mechanisms such as Dynamic Client Registration [@!RFC7591] and OpenID Federation [@!OpenID.Federation] are still applicable.
 
-Header
+To allow In-band Registration mechanism as well as existing mechanisms to be supported in parallel, Wallets and Wallet Providers MUST enforce sufficient randomness of the `client_id` for pre-registered Clients.
 
-<{{examples/request/request_header_client_id_did.json}}
+Wallets MUST verify the integrity of the `client_id`, `redirect_uri` and/or `response_uri` parameter from the request. Additionally, Wallets SHOULD verify the `redirect_uri` and/or `response_uri` belong to the Client with the identified by the `client_id`.
 
-Body
+This specification defines the following mechanisms for In-band Registration for signed requests:
+* X.509 Certificates: If supported by the Wallet, the request is signed, and any of the `x5c`, `x5t`, or `x5t#S256` JWT headers are present, the Wallet MUST validate the corresponding X.509 certificate chain and use the public key from the leaf certificate to verify the signature of the signed request to authenticate the request. Additionally, the Wallet MUST check that the `client_id` in the signed request matches any X.509 Subject Alternative Name (SAN) value of the leaf certificate. If the Wallet can establish trust in the `client_id` authenticated through the leaf certificate, e.g., because the `client_id` is contained in a list of trusted Client Identifiers, it MAY allow the Client to freely choose the `redirect_uri` or `response_uri` value. If not, the `redirect_uri` or `response_uri` value MUST match the `client_id` or FQDN of the `client_id`. If either verification fails, the Wallet MUST reject the request, otherwise processing continues.
+* Verifier Attestations: If supported by the Walletm the request is signed, and the signed request contains a `jwt` JWT header that is a Verifier Attestation JWT, the Wallet MUST use the public key in the `cnf` claim in the Verifier Attestation JWT to verify the signatuer of the signed request. This serves as proof of possession of this key. Additionally, the `client_id` MUST equal the `sub` claim value in the Verifier Attestation JWT. The Wallet MUST validate the signature on the Verifier Attestation JWT. The `iss` claim value of the Verifier Attestation JWT MUST identify a party the Wallet trusts for issuing Verifier Attestation JWTs. If the issuer of the Verifier Attestation JWT includes a `redirect_uris` claim in the Verifier Attestation, the Wallet MUST verify that the `redirect_uri` or `response_uri` request parameter value exactly matches one of the `redirect_uris` array values. If either of the checks fail, the Wallet MUST reject the request, otherwise processing continues.
```

Comment:
```suggestion
* Verifier Attestations: If supported by the Wallet, the request is signed, and the signed request contains a `jwt` JWT header that is a Verifier Attestation JWT, the Wallet MUST use the public key in the `cnf` claim in the Verifier Attestation JWT to verify the signatuer of the signed request. This serves as proof of possession of this key. Additionally, the `client_id` MUST equal the `sub` claim value in the Verifier Attestation JWT. The Wallet MUST validate the signature on the Verifier Attestation JWT. The `iss` claim value of the Verifier Attestation JWT MUST identify a party the Wallet trusts for issuing Verifier Attestation JWTs. If the issuer of the Verifier Attestation JWT includes a `redirect_uris` claim in the Verifier Attestation, the Wallet MUST verify that the `redirect_uri` or `response_uri` request parameter value exactly matches one of the `redirect_uris` array values. If either of the checks fail, the Wallet MUST reject the request, otherwise processing continues.
```